x/emissions/metrics/metrics_util.go

```diff
@@ -7,53 +7,103 @@ import (
 	"time"
 
 	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
 	metrics "github.com/hashicorp/go-metrics"
 )
 
 // Measures RPC endpoint request throughput
 // Metric Name:
 //
 //	allora_rpc_request_counter
-func IncrementRpcRequestCounter(endpoint string, err *error) {
+func IncrementRpcRequestCounter(endpoint string, err *error, labels map[string]string) {
 	success := *err == nil
+	metricLabels := []metrics.Label{
+		telemetry.NewLabel("endpoint", endpoint),
+		telemetry.NewLabel("success", strconv.FormatBool(success)),
+	}
+
+	// Add any additional labels
+	for k, v := range labels {
+		metricLabels = append(metricLabels, telemetry.NewLabel(k, v))
+	}
+
 	telemetry.IncrCounterWithLabels(
 		[]string{"allora", "request", "counter"},
 		float32(1),
-		[]metrics.Label{
-			telemetry.NewLabel("endpoint", endpoint),
-			telemetry.NewLabel("success", strconv.FormatBool(success)),
-		},
+		metricLabels,
 	)
 }
 
 // Measures the RPC request latency in milliseconds
 // Metric Name:
 //
-//	allora_rpc_request_latency_ms
-func MeasureRpcRequestLatency(endpoint string, startTime time.Time) {
+//	allora_request_latency_ms
+func MeasureRpcRequestLatency(endpoint string, startTime time.Time, labels map[string]string) {
+	allLabels := []metrics.Label{telemetry.NewLabel("endpoint", endpoint)}
+	for k, v := range labels {
+		allLabels = append(allLabels, telemetry.NewLabel(k, v))
+	}
 	metrics.MeasureSinceWithLabels(
 		[]string{"allora", "request", "latency_ms"},
 		startTime.UTC(),
-		[]metrics.Label{
-			telemetry.NewLabel("endpoint", endpoint),
-		},
+		allLabels,
 	)
 }
 
-// IncrProducerEventCount increments the counter for events produced.
-// This metric counts the number of events produced by the system.
+// IncrProducerEventCountWithLabels increments the counter for events produced, with custom labels.
+// This metric counts the number of events produced by the system, with additional labels.
 // Metric Name:
 //
 //	allora_loadtest_produce_count
-func IncrProducerEventCount(msgType string) {
+func IncrProducerEventCountWithLabels(msgType string, labels map[string]string) {
+	// Always include the msg_type label
+	allLabels := []metrics.Label{telemetry.NewLabel("msg_type", msgType)}
+	for k, v := range labels {
+		allLabels = append(allLabels, telemetry.NewLabel(k, v))
+	}
 	telemetry.IncrCounterWithLabels(
 		[]string{"allora", "loadtest", "produce", "count"},
 		1,
-		[]metrics.Label{telemetry.NewLabel("msg_type", msgType)},
+		allLabels,
 	)
 }
 
-func RecordMetrics(apiMethod string, startTime time.Time, err *error) {
-	IncrementRpcRequestCounter(apiMethod, err)
-	MeasureRpcRequestLatency(apiMethod, startTime)
+func RecordMetrics(apiMethod string, startTime time.Time, err *error, labels map[string]string) {
+	if labels == nil {
+		labels = make(map[string]string)
+	}
```

Comment:
```suggestion
	if labels == nil {
		labels = make(map[string]string)
	}
```
We can read from a `nil` map so let it be `nil`